pkg/controller/walle/controller.go

```diff
@@ -0,0 +1,443 @@
+			case <-wr.stopCh:
+				informer.close()
+				return
+			default:
+				informer.run()
+			}
+		}
+	}(informer)
+
+	return nil
+}
+
+func (wr *WalleController) klusterAddFunc(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err == nil {
+		wr.queue.Add(key)
+	}
+}
+
+func (wr *WalleController) klusterUpdateFunc(cur, old interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(cur)
+	if err == nil {
+		wr.queue.Add(key)
+	}
+}
+
+func (wr *WalleController) klusterDeleteFunc(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err == nil {
+		wr.queue.Add(key)
+	}
+	obj, exists, _ := wr.Factories.Kubernikus.Kubernikus().V1().Klusters().Informer().GetIndexer().GetByKey(key)
+	if exists {
+		obj.(*NodeInformer).close()
+		wr.nodeInformerMap.Delete(key)
+	}
+}
+
+func revertKey(key string) (string, string) {
+	s := strings.Split(key, "/")
+	if len(s) == 2 {
+		return s[0], s[1]
+	}
+	return "", ""
+}
+
+// GetStore returns the SharedIndexInformers cache.Store for a given kluster or an error
+func (wr *WalleController) GetStore(kluster *v1.Kluster) (cache.Store, error) {
+	informer, err := wr.getNodeInformerByKluster(kluster)
+	if err != nil {
+		return nil, err
+	}
+	return informer.GetStore(), nil
+}
+
+// GetIndexer returns the SharedIndexInformers cache.Indexer for a given kluster or an error
+func (wr *WalleController) GetIndexer(kluster *v1.Kluster) (cache.Indexer, error) {
+	informer, err := wr.getNodeInformerByKluster(kluster)
+	if err != nil {
+		return nil, err
+	}
+	return informer.GetIndexer(), nil
+}
+
+func (wr *WalleController) getNodeInformerByKluster(kluster *v1.Kluster) (NodeInformer, error) {
+	key, err := cache.MetaNamespaceKeyFunc(kluster)
+	if err != nil {
+		return NodeInformer{}, err
+	}
+	informer, ok := wr.nodeInformerMap.Load(key)
+	if ok {
+		return informer.(NodeInformer), nil
+	}
+	return NodeInformer{}, fmt.Errorf("no informer found for kluster %v", key)
+}
+
+// List..
+func (wr *WalleController) List() map[*v1.Kluster]cache.SharedIndexInformer {
+	m := make(map[*v1.Kluster]cache.SharedIndexInformer)
+	wr.nodeInformerMap.Range(
+		func(key, value interface{}) bool {
+			ns, name := revertKey(key.(string))
+			k, err := wr.Clients.Kubernikus.KubernikusV1().Klusters(ns).Get(name, meta_v1.GetOptions{})
+			if err != nil {
+				wr.logger.Log("Couldn't get kluster %s/%s while listing all", ns, name)
+			}
+			m[k] = value.(NodeInformer).SharedIndexInformer
+			return true
+		},
+	)
+	return m
+}
+
+func (wr *WalleController) HasSynced(kluster *v1.Kluster) bool {
+	key, err := cache.MetaNamespaceKeyFunc(kluster)
+	if err != nil {
+		return false
+	}
+	informer, ok := wr.nodeInformerMap.Load(key)
+	if ok {
+		return false
+	}
+	return informer.(NodeInformer).HasSynced()
+}
+
+// AddEventHandlerAddFunc ...
+func (wr *WalleController) addEventHandlerAddFunc(addFunc func(node *api_v1.Node)) {
+	if addFunc != nil {
+		wr.eventHandlerFuncs.Store(funcType.AddFunc, addFunc)
```

Comment:
I think this is not working as intended as only one handler is stored for a specific type (add, update, delete). It should be possible to register more then one handler that is invoked on a node event.